### ISSUE TITLE
Missing trailing expression in block with `let else` yields incorrectly spanned error message

### ISSUE THREAD

**Issue by JakobDegen** [rust-lang/rust#94176]
I tried this code:

```rust
#![feature(let_else)]

pub fn test(a: Option<u32>) -> Option<u32> {
    let Some(_) = a else { return None; };
    println!("Foo");
}

fn main() {}
```

I expected to see this happen: A helpful error message pointing to the end of the block.

Instead, this happened: An error message with a wrong span:

```rust
error[E0308]: mismatched types
 --> test.rs:4:5
  |
4 |     let Some(_) = a else { return None; };
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Option`, found `()`
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```

### Meta

`rustc --version --verbose`:
```
rustc 1.60.0-nightly (b17226fcc 2022-02-18)
binary: rustc
commit-hash: b17226fcc11587fed612631be372a5b4cb89988a
commit-date: 2022-02-18
host: x86_64-unknown-linux-gnu
release: 1.60.0-nightly
LLVM version: 14.0.0
```

cc #87335

@rustbot labels: +T-compiler +F-let-else +requires-nightly +A-diagnostics

**Comment by JakobDegen** [comment]
Let's try that again...

@rustbot label +T-compiler +F-let-else +requires-nightly +A-diagnostics

**Comment by est31** [comment]
FTR this is the error for the `if let` equivalent:

```Rust
pub fn test(a: Option<u32>) -> Option<u32> {
    if let Some(_) = a {
    } else {
        return None;
    }
    println!("Foo");
}
```

```
error[E0308]: mismatched types
 --> src/main.rs:3:32
  |
3 | pub fn test(a: Option<u32>) -> Option<u32> {
  |        ----                    ^^^^^^^^^^^ expected enum `Option`, found `()`
  |        |
  |        implicitly returns `()` as its body has no tail or `return` expression
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`
```

**Comment by est31** [comment]
I wonder how #94012 is affecting this, as this seems to be a bug in the desugaring and that PR changes it. @cormacrelf what is the error output with your PR applied? Does it fix the issue? If so, maybe adding a test to your PR would be helpful.

**Comment by est31** [comment]
I meant *caused by* the desugaring, not *in* the desugaring.

**Comment by cormacrelf** [comment]
It won't fix it. This needs a bit of work I think, to push the error further inside the if-let-else branch, so the span marked is the implicit return expression.

**Comment by est31** [comment]
Hmm if you move the println into the `if let`, it gives a similar error to the one emitted by `let else`:

```Rust
pub fn test(a: Option<u32>) -> Option<u32> {
    if let Some(_) = a {
        println!("Foo");
    } else {
        return None;
    }
}
```
gives
```
error[E0308]: mismatched types
 --> src/main.rs:2:24
  |
2 |       if let Some(_) = a {
  |  ________________________^
3 | |         println!("Foo");
4 | |     } else {
  | |_____^ expected enum `Option`, found `()`
  |
  = note:   expected enum `Option<u32>`
          found unit type `()`
```

**Comment by cormacrelf** [comment]
That's because you have only now reproduced the actual desugaring.

**Comment by cormacrelf** [comment]
I'm starting to think the nested desugaring is not such a good idea. Even the guy who implemented it forgot how it worked the other day, it continues to mystify, and now we have some more error message complications. It's less work in hir lowering, but has absolutely no other benefits.
